File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.java
import java.util.*;

class Solution {
    public boolean[] visited;
    
    public void dfs(List<ArrayList<Integer>> graph, int vertex) {
        if (visited[vertex]) {
            return;
        }
        
        visited[vertex] = true;
        
        for (Integer next : graph.get(vertex)) {
            dfs(graph, next);
        }
    }
    
    public boolean validPath(int n, int[][] edges, int source, int destination) {
        List<ArrayList<Integer>> graph = new ArrayList<>();
        visited = new boolean[n];    
        
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        
        for (int i = 0; i < edges.length; i++) {
            graph.get(edges[i][0]).add(edges[i][1]);
            graph.get(edges[i][1]).add(edges[i][0]);
        }    
        
        dfs(graph, source);
        return visited[destination];
    }
}

File: 1971-find-if-path-exists-in-graph/1971-find-if-path-exists-in-graph.cpp
class Solution {
public:
  typedef vector<int>vi;
  typedef vector<vi>vii;
  typedef vector<bool>vb;
  vii graph;
  int s,d;
  vb visited;
  bool dfs(int cur){
    if(visited[cur]){
      return false;
    }
    if(cur==d){
      return true;
    }
    visited[cur]=true;
    for(auto next:graph[cur]){
      if(dfs(next)){
        return true;
      }
    }
    return false;
  }
  bool validPath(int n, vector<vector<int>>& edges, int source, int destination) {
    s=source;d=destination;
    visited=vb(n,false);
    graph=vii(n);
    for(int i=0;i<edges.size();i++){
      graph[edges[i][0]].push_back(edges[i][1]);
      graph[edges[i][1]].push_back(edges[i][0]);
    }
    return dfs(s);
  }
};